janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java

```diff
@@ -85,48 +87,79 @@ public HttpActionService(String baseUrl, HttpClient client, Converter converter)
-                private int lastProgress;
-
-                @Override public void onProgress(int progress) {
-                    if (progress > lastProgress + PROGRESS_THRESHOLD) {
-                        callback.onProgress(holder, progress);
-                        lastProgress = progress;
+            request = builder.build();
+            putRunningRequest(action, request);
+            throwIfCanceled(action, request);
+            try {
+                response = client.execute(request, new ActionRequestCallback<A>(holder) {
+                    private int lastProgress;
+
+                    @Override public void onProgress(int progress) {
+                        if (progress > lastProgress + PROGRESS_THRESHOLD) {
+                            callback.onProgress(holder, progress);
+                            lastProgress = progress;
+                        }
                     }
-                }
-            });
-            throwIfCanceled(action);
+                });
+            } finally {
+                throwIfCanceled(action, request);
+            }
             if (!response.isSuccessful()) {
                 throw new HttpException(response.getStatus(), response.getReason());
             }
             action = helper.onResponse(action, response, converter);
+            throwIfCanceled(action, request);
         } catch (CancelException e) {
             return;
         } catch (Throwable e) {
             throw new HttpServiceException(e);
         } finally {
-            runningActions.remove(action);
+            if (request != null) {
+                List<Request> requests = runningRequests.get(action);
+                if (requests != null) {
+                    requests.remove(request);
+                }
+            }
         }
         this.callback.onSuccess(holder);
     }
 
     @Override protected <A> void cancel(ActionHolder<A> holder) {
-        runningActions.remove(holder.action());
+        A action = holder.action();
+        List<Request> requests = runningRequests.remove(action);
+        try {
+            if (requests != null) {
+                for (Request request : requests) {
+                    client.cancel(request);
+                }
+            }
+        } catch (Throwable ignored) {}
+    }
+
+    private void putRunningRequest(Object action, Request request) {
+        putRunningAction(action);
+        runningRequests.get(action).add(request);
+    }
+
+    private void putRunningAction(Object action) {
+        if (!runningRequests.containsKey(action)) {
+            runningRequests.put(action, new CopyOnWriteArrayList<Request>());
+        }
     }
 
-    private void throwIfCanceled(Object action) throws CancelException {
-        if (!runningActions.contains(action)) {
+    private void throwIfCanceled(Object action, Request request) throws CancelException {
+        if (!runningRequests.containsKey(action)
```

Comment:
containsKey and get are equal operations they do literally the same searching algorithm. So it's obsolete for perf. Should use local var instead?